### ISSUE TITLE
Debugging Malformed MELPA Recipe: flymake-collection

### ISSUE THREAD

**Issue by mohkale** [raxod502/straight.el#941]
Hi, I'm having issues with a recipe I recently published to MELPA called [flymake-collection](https://github.com/melpa/melpa/pull/7988). When my emacs starts up it errors saying it couldn't find a recipe for `flymake-collection`. After startup when I do `M-x straight-pull-recipe-repositories` followed by `M-x straight-use-package flymake-collection` it complains about a malformed recipe. Could you help me debug what could be causing this?

For reference I have this in my `init.el`:

```elisp
(use-package flymake-collection :straight t)
```

**Comment by progfolio** [comment]
Unfortunately, I'm unable to reproduce your error in a clean environment:

[Yodel](https://github.com/progfolio/yodel) Report (2022-04-18 12:32:51):

```emacs-lisp
(yodel
  :save "flymake-collection-test"
  :packages*
  flymake-collection
  :post*
  (message "%S"
           (locate-library "flymake-collection")))
```

<details><summary>STDOUT:</summary>

```emacs-lisp
Loading /tmp/flymake-collection-test/straight-bootstrap-snippet.el (source)...
Bootstrapping straight.el...
Bootstrapping straight.el...done
Rebuilding all packages due to build cache schema change
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done
Cloning flymake-collection...
Cloning flymake-collection...done
Building flymake-collection...
Building flymake-collection → Cloning let-alist...
Building flymake-collection → Cloning let-alist...done
Building flymake-collection → Building let-alist...
Building flymake-collection → Building let-alist...done
Building flymake-collection → Cloning flymake...
Building flymake-collection → Cloning flymake...done
Building flymake-collection → Building flymake...
Building flymake-collection → Building flymake → Cloning eldoc...
Building flymake-collection → Building flymake → Cloning eldoc...done
Building flymake-collection → Building flymake → Building eldoc...
Building flymake-collection → Building flymake → Building eldoc...done
Building flymake-collection → Building flymake → Cloning project...
Building flymake-collection → Building flymake → Cloning project...done
Building flymake-collection → Building flymake → Building project...
Building flymake-collection → Building flymake → Building project → Cloning xref...
Building flymake-collection → Building flymake → Building project → Cloning xref...done
Building flymake-collection → Building flymake → Building project → Building xref...
Building flymake-collection → Building flymake → Building project → Building xref...done
Building flymake-collection → Building flymake → Building project...
Building flymake-collection → Building flymake → Building project...done
Building flymake-collection → Building flymake...
Building flymake-collection → Building flymake...done
Building flymake-collection...
Building flymake-collection...done

"/tmp/flymake-collection-test/straight/build/flymake-collection/flymake-collection.elc"
```

</details>
<details><summary>Environment</summary>

- **emacs version**: GNU Emacs 29.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.33, cairo version 1.17.6)
 of 2022-04-12
- **system type**: gnu/linux

<details open><summary>Packages</summary>

| Name                                                                | Branch  | Commit                                                                                        | Date       | Source          |
|---------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------|------------|-----------------|
| [flymake-collection](https://github.com/mohkale/flymake-collection) | release | https://github.com/mohkale/flymake-collection/commit/297406f601daffbec2b3305013442b2829451416 | 2022-04-12 | melpa           |
| [let-alist](https://github.com/emacs-straight/let-alist)            | master  | https://github.com/emacs-straight/let-alist/commit/592553db5929b54db40af0df90c5add0aaca045b   | 2022-01-02 | gnu-elpa-mirror |
| [flymake](https://github.com/emacs-straight/flymake)                | master  | https://github.com/emacs-straight/flymake/commit/46e3682127c183e4a58601bcba18a67e6bb27912     | 2022-02-11 | gnu-elpa-mirror |
| [eldoc](https://github.com/emacs-straight/eldoc)                    | master  | https://github.com/emacs-straight/eldoc/commit/eab3f2590621c6559cb92a5edc519fc7e51ef850       | 2022-03-25 | gnu-elpa-mirror |
| [project](https://github.com/emacs-straight/project)                | master  | https://github.com/emacs-straight/project/commit/cad9133f15a04e2b55b0f4f8ce1d3e35d946ebe6     | 2022-04-17 | gnu-elpa-mirror |
| [xref](https://github.com/emacs-straight/xref)                      | master  | https://github.com/emacs-straight/xref/commit/17201455d06025a095e7db776e31d4229020908d        | 2022-04-18 | gnu-elpa-mirror |

</details>

</details>


Could you try evaluating the following test case in your scratch buffer?:

```emacs-lisp
(straight-bug-report
  :post-bootstrap
  (straight-use-package 'flymake-collection)
  (message "%S" (locate-library "flymake-collection")))
```

I apologize for the informative "malformed recipe" error.
It's something we could definitely improve with straight.el.
However, the recipe looks fine to me (straight doesn't currently have any support for the :old-names recipe keyword, but that should not prevent the new recipe from working).
My initial hunch is that perhaps the cached version of the recipe is malformed in some way.
What is the output of the following?:

```emacs-lisp
(straight-recipes-retrieve 'flymake-collection) 
```

**Comment by mohkale** [comment]
@progfolio 

> (straight-recipes-retrieve 'flymake-collection) 

This returns nil.

> straight-bug-report


<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
    :post-bootstrap 
    (straight-use-package 'flymake-collection)
    (message "%S"
        (locate-library "flymake-collection")))
```
</details>

- Test run at: `2022-04-19 19:46:35`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> master, origin/master, origin/HEAD) af5437f 2021-09-27`
- emacs-version: `GNU Emacs 29.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.30, cairo version 1.17.4)
 of 2021-10-22`

<details><summary>Output</summary>

```emacs-lisp
Building flymake-collection → Building flymake → Building project → Building xref...done
Building flymake-collection → Building flymake → Building project...
Building flymake-collection → Building flymake → Building project...done
Building flymake-collection → Building flymake...
Building flymake-collection → Building flymake...done
Building flymake-collection...
Building flymake-collection...done

"/tmp/straight.el-test-mNOe7d/straight/build/flymake-collection/flymake-collection.elc"
Packages:
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master 12505571 2022-04-17
"gnu-elpa-mirror"         n/a                  master e9835ab 2022-04-16
"el-get"                  melpa                master 93533097 2021-12-24
"emacsmirror-mirror"      n/a                  master 25f12aa 2022-04-13
"straight"                n/a                  develop 4517e11 2022-04-13
"flymake-collection"      melpa                release 297406f 2022-04-12
"let-alist"               gnu-elpa-mirror      master 592553d 2022-01-02
"flymake"                 gnu-elpa-mirror      master 46e3682 2022-02-11
"eldoc"                   gnu-elpa-mirror      master eab3f25 2022-03-25
"project"                 gnu-elpa-mirror      master cad9133 2022-04-17
"xref"                    gnu-elpa-mirror      master 1720145 2022-04-18

<!-- copy entire buffer output and paste in an issue at:
https://github.com/raxod502/straight.el/issues/new/choose -->
<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
    :post-bootstrap 
    (straight-use-package 'flymake-collection)
    (message "%S"
        (locate-library "flymake-collection")))
```
</details>

- Test run at: `2022-04-19 19:46:33`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> master, origin/master, origin/HEAD) af5437f 2021-09-27`
- emacs-version: `GNU Emacs 29.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.30, cairo version 1.17.4)
 of 2021-10-22`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Rebuilding all packages due to build cache schema change
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) 4517e11 2022-04-13
Cloning flymake-collection...
Building straight...done

Test run with version: prerelease (HEAD -> develop, origin/develop) 4517e11 2022-04-13
Cloning flymake-collection...
Cloning flymake-collection...done
Building flymake-collection...
Building flymake-collection → Cloning let-alist...
Cloning flymake-collection...done
Building flymake-collection...
Building flymake-collection → Cloning let-alist...
Building flymake-collection → Cloning let-alist...done
Building flymake-collection → Building let-alist...
Building flymake-collection → Building let-alist...done
Building flymake-collection → Cloning flymake...
Building flymake-collection → Cloning let-alist...done
Building flymake-collection → Building let-alist...
Building flymake-collection → Building let-alist...done
Building flymake-collection → Cloning flymake...
Building flymake-collection → Cloning flymake...done
Building flymake-collection → Building flymake...
Building flymake-collection → Building flymake → Cloning eldoc...
Building flymake-collection → Cloning flymake...done
Building flymake-collection → Building flymake...
Building flymake-collection → Building flymake → Cloning eldoc...
Building flymake-collection → Building flymake → Cloning eldoc...done
Building flymake-collection → Building flymake → Building eldoc...
Building flymake-collection → Building flymake → Building eldoc...done
Building flymake-collection → Building flymake → Cloning project...
Building flymake-collection → Building flymake → Cloning eldoc...done
Building flymake-collection → Building flymake → Building eldoc...
Building flymake-collection → Building flymake → Building eldoc...done
Building flymake-collection → Building flymake → Cloning project...
Building flymake-collection → Building flymake → Cloning project...done
Building flymake-collection → Building flymake → Building project...
Building flymake-collection → Building flymake → Building project → Cloning xref...
Building flymake-collection → Building flymake → Building project → Cloning xref...done
Building flymake-collection → Building flymake → Building project → Building xref...
Building flymake-collection → Building flymake → Building project → Building xref...done
Building flymake-collection → Building flymake → Building project...
Building flymake-collection → Building flymake → Cloning project...done
Building flymake-collection → Building flymake → Building project...
Building flymake-collection → Building flymake → Building project → Cloning xref...
Building flymake-collection → Building flymake → Building project...done
Building flymake-collection → Building flymake...
Building flymake-collection → Building flymake...done
Building flymake-collection...
Building flymake-collection → Building flymake → Building project → Cloning xref...done
Building flymake-collection → Building flymake → Building project → Building xref...
Building flymake-collection...done

"/tmp/straight.el-test-AAyLQX/straight/build/flymake-collection/flymake-collection.elc"
Packages:
"org-elpa"                n/a                  n/a
"melpa"                   n/a                  master 12505571 2022-04-17
"gnu-elpa-mirror"         n/a                  master e9835ab 2022-04-16
"el-get"                  melpa                master 93533097 2021-12-24
"emacsmirror-mirror"      n/a                  master 25f12aa 2022-04-13
"straight"                n/a                  develop 4517e11 2022-04-13
"flymake-collection"      melpa                release 297406f 2022-04-12
"let-alist"               gnu-elpa-mirror      master 592553d 2022-01-02
"flymake"                 gnu-elpa-mirror      master 46e3682 2022-02-11
"eldoc"                   gnu-elpa-mirror      master eab3f25 2022-03-25
"project"                 gnu-elpa-mirror      master cad9133 2022-04-17
"xref"                    gnu-elpa-mirror      master 1720145 2022-04-18


```
</details>
```
</details>

Funnily enough I ran this in an emacs which failed at startup because it couldn't install the package, yet the test passed :rofl:.

> My initial hunch is that perhaps the cached version of the recipe is malformed in some way.

I suspected as much, and I did see a bunch of cache variables in straight.el while looking for what might be causing this but `M-x straight cache` (I use orderless) didn't show any possible commands (where I expected a command like `straight-clear-cache`) so I didn't think it would be related. I also though running `straight-pull-recipe-repositories` would invalidate the cache anyways and I also tried just renaming my root straight install directory so straight tried to install everything from scratch, but that still failed :-(.

**Comment by progfolio** [comment]
> Funnily enough I ran this in an emacs which failed at startup because it couldn't install the package, yet the test passed rofl.

That's a good sign. The test is run in a clean environment. So it looks like we don't have a bug in straight.el here.

> 

    (straight-recipes-retrieve 'flymake-collection)

This returns nil.

That likely means the MELPA repository is out of date.
You can manually check via git in the repository to see if you're on a commit which contains your package's new recipe.


> straight-version: prerelease (HEAD -> master, origin/master, origin/HEAD) af5437f 2021-09-27

I recommend upgrading to the develop branch and seeing if that helps. You can do so by adding:

```emacs-lisp
(setq straight-repository-branch "develop")
```
prior to the bootstrapping snippet for straight.el in your init file.

**Comment by mohkale** [comment]
> That likely means the MELPA repository is out of date.
> You can manually check via git in the repository to see if you're on a commit which contains your package's new recipe.

I manually pulled the MELPA repository and checked to ensure it was up to date but I'm still encountering the issue :-(.

> I recommend upgrading to the develop branch and seeing if that helps.

Didn't work either sadly.

Do you know where the cache is being saved? I'm encountering this even after deleting my entire straight directory. If it's working in the test I'm guessing the cache doesn't apply there, so how do I clear/invalidate the cache?

**Comment by progfolio** [comment]
> Do you know where the cache is being saved? I'm encountering this even after deleting my entire straight directory.

under the straight directory there is a `build-cache.el` file which can be removed. However, 

>  I'm encountering this even after deleting my entire straight directory. 

indicates that the issue probably isn't with the cache.

If you have your init.el uploaded somewhere, I'd be happy to take a look to see if anything looks off in that regard.

**Comment by mohkale** [comment]
@progfolio 

You can find my config [here](https://github.com/mohkale/emacs). I recommend cloning, running `./bin/org-tangle $(pwd)/init.org` and then loading init.el.

**Comment by progfolio** [comment]
Unfortunately, I'm not able to get your config to launch do to an entirely different error. At a glance, though, it looks like you're advising some parts of straight.el.

Considering the recipe works fine in a clean test environment, my hunch is that the problem is somewhere in your init. You could always try adding a `(debug)` statement just before the declaration for flymake-collection. Launching emacs with the `--debug-init` flag will allow you to step through the processing of that package and may give some insight to where the issue lies.

Let me know if you find anything we can act on in straight.el.

**Comment by mohkale** [comment]
@progfolio 

Glad to say the issue seams to have fixed itself. In regards to the advice you mentioned that's setup so that the straight lockfile is in my emacs directory and not my straight directory because I place straight under XDG_CACHE_HOME, and not my .emacs.d directory. It looks like when I deleted straight/ the lockfile persisted and pulled in an older version of MELPA from the version pinned in my Lockfile the next time I ran emacs. I'm still not sure why manually pulling the recipe repositories doesn't fix the issue :/. but either way it's working for me so this issue can be closed.
